### DISCUSSION TITLE
Speedrun Optimization Writeup

### DISCUSSION THREAD

**Discussion by evan-conway** [KellerJordan/modded-nanogpt#211] [General]
Hi, I'm interested in doing an explanatory writeup on some of the optimizations presented here, and wanted to reach out to see if there’s any feedback on what kind of things should be included.

**Why:** When looking at the list of records in the repository, it can be hard to tell which optimizations gave the greatest improvements, and how to implement those optimizations. This is made particularly difficult by the fact that many of the optimization details are scattered across various PRs and X threads.

**Goal:** Produce a writeup detailing the major sources of speedups and providing citations for how those optimizations were derived. Ideally, people should be able to read through this and get ideas for making their own model faster.

As I currently see it, there are a few major initial sources of improvement to cover. Together, these should give a speedrun time of around 3.5 minutes. For more details, I have a rough spreadsheet of optimizations [here](https://docs.google.com/spreadsheets/d/1ocJtEM9FtnRfN8jT0pYvcQtD237ns4eq-mcOXWfg7hM/edit?usp=sharing).
1. **Optimized Muon (77.53% speedup):** Muon optimizer with [initial improvements](https://x.com/kellerjordan0/status/1844820919061287009), distributed overhead, and momentum warmup
   1. I’m ignoring most of the later Muon improvements, like NorMuon, since they seem to make less of an individual impact. However, some might be worth including if they simplify things
2. **FlexAttention (54.51% speedup):** switching to FlexAttention with block size warmup
3. **Initial Tuning (46.05% speedup):** tuning learning rate and padding embeddings to a power of two
4. **Architecture Modernization (45.21% speedup):** dropping in architecture modifications that are primarily taken from existing literature without too many changes. More specifically, this consists of adding RoPE, QK-norm, zero-initialized projections, and ReLU^2 activation
5. **Skip Connections with Value Embeddings (36.46% speedup):** U-Net style skip connections, coupled with value embeddings
   1. Later, the skip connections are simplified in https://github.com/KellerJordan/modded-nanogpt/pull/159, leaving only one skip connection if I understand correctly
6. **Logit Softcap (20.86% speedup):** softcapping logits at 15, with offset logits and asymmetric rescaling
7. **PyTorch Updates (11.81% speedup)**
8. **Untied Embeddings (11.11% speedup):** untying embeddings and head
   1. This is later retied for another gain, and then switched to tied for 2/3 of training for yet another gain. I’m not completely sure how to interpret which of these options works best with the limited set of optimizations outlined here. Simplest option would be not to untie at all, since that’s the default here
9. **Lower Precision (8.03% speedup):** bfloat16 activations and a tuned fp8 head

Some other possible things to include:

- Bigram Hash Embeddings
- Batch Size Improvements
- Further Muon Optimizations
- Paired Head Attention

Does this selection look good? I'm interested in hearing what kind of optimizations to include, how they should be grouped together, what details are most important (theory vs code), along with anything else you want to mention.

**Comment by ClassicLarry** [comment]
Here is an Oct blog post that focused on the July-October changes: https://www.lesswrong.com/posts/j3gp8tebQiFJqzBgg/how-the-nanogpt-speedrun-wr-dropped-by-20-in-3-months#fn6qk7h03qtp5.

And a Nov one on Muon implementation: https://varunneal.github.io/essays/muon

Overall your list looks good. Flex attention was replaced with flash attention 3. Also I want to emphasis that these speed-ups are optimized for this particular scale. If we were hyper-optimizing for the 10B scale we would do different approaches.

**Reply by evan-conway** [reply]
Thanks for the response and the notes! I had seen your LessWrong post; that's actually part of what got me interested in doing a writeup on this, since I was curious about some of the earlier optimizations. However, I hadn't seen that post on Muon.

With regard to the model size, I was actually curious if there was any information about which improvements are scalable, since I had wanted to include that after seeing that the version transferred to the GPT-2 speedrun did quite well. I know that Muon at least has had good results at larger sizes, but don't know much about the others.

**Comment by evan-conway** [comment]
Writeup is accessible here, I delayed it a bit so I could include some of the recent progress from NanoChat: https://evanjayconway.com/posts/2026/nanogpt-improvements/

Please let me know if you have any feedback!
